delft3dworker/models.py

```diff
@@ -479,11 +490,160 @@ class Container(models.Model):
     # docker container ids are sha256 hashes
     docker_id = models.CharField(max_length=64, blank=True, default='')
 
-    def _update_state_and_save(self):
+    docker_log = models.TextField(blank=True, default='')
 
-        # TODO: write _update_state_and_save method
+    def update_task_result(self):
+        """
+        This method will get the result from the last task it executed, given
+        that there is a result. If the task is not ready, this will not do
+        anything.
+        """
+
+        if self.task_uuid is None:
+            return
+
+        result = AsyncResult(id=self.task_uuid)
+
+        if result.ready():
+
+            self.docker_id, self.docker_log = result.get()
+            self.task_uuid = None
+            self.save()
+
+    def update_from_docker_snapshot(self, snapshot):
+        """
+        This method updates the Container based on a given snapshot
+        of a docker container which was retrieved with docker-py's
+        client.containers(all=True) (equivalent to 'docker ps').
+
+        Given that the container has no pending tasks, it will also compare
+        this state to the its desired_state, which is defined by the Scene to
+        which this Container belongs. If (for any reason) the docker_state is
+        different from the desired_state, this Container will act: it will
+        start a task to get both states matched.
+
+        At the end the Container will request a log update.
+        """
+
+        self._update_state_and_save(snapshot)
+
+        if self.task_uuid == '':
+
+            self._fix_state_mismatch()
+
+            self._update_log()
+
+    def _update_state_and_save(self, snapshot):
+        """
+        Var snapshot can be either dictionary or None.
+        If None: docker container does not exist
+        If dictionary: snapshot['Status'] is a string describing status
+        """
+
+        if snapshot is None:
+            self.docker_state = 'non-existent'
+
+        elif isinstance(snapshot, dict) and ('Status' in snapshot):
+
+            if snapshot['Status'].startswith('Up'):
+                self.docker_state = 'running'
+
+            elif snapshot['Status'].startswith('Created'):
+                self.docker_state = 'created'
+
+            elif snapshot['Status'].startswith('Exited'):
+                self.docker_state = 'exited'
+
+            else:
+                self.docker_state = 'unknown'
+
+        else:
```

Comment:
Can we log an error here (if unknown)
Snapshot is either a dict with a matchable Id key, or None. Thus, either missing a 'Status' or not known state indicates an existing docker output we cannot parse.